llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp

```diff
@@ -0,0 +1,160 @@
+//===-- AIEEliminateDuplicatePHI.cpp - Identify and remove duplicate PHI --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+// Identify and replace duplicate PHI nodes uses with the base PHI node.
+// The pass specifically targets PHI nodes with pointer type operands.
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseSubtarget.h"
+#include "llvm/CodeGen/GlobalISel/CSEInfo.h"
+#include "llvm/CodeGen/GlobalISel/CSEMIRBuilder.h"
+#include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+
+#define DEBUG_TYPE "aie-duplicate-PHI-elimination"
+
+using namespace llvm;
+
+static const char AIE_ELIMINATE_DUPLICATE_PHI[] =
+    "AIE Eliminate Duplicate PHI Pass";
+
+namespace {
+
+/// Eliminates duplicate PHI uses by replacing all uses of DuplicateReg with
+/// BaseReg.
+///
+/// This function iterates over all instructions that use DuplicateReg and
+/// replaces any register operands that match DuplicateReg with BaseReg.
+///
+/// \param MRI The MachineRegisterInfo object for the current function.
+/// \param Observer GISelObserverWrapper object to observe and track changes.
+/// \param BaseReg The register to replace DuplicateReg with.
+/// \param DuplicateReg The register to be replaced.
+static void eliminateDuplicatePHIUses(MachineRegisterInfo &MRI,
+                                      GISelObserverWrapper &Observer,
+                                      Register BaseReg, Register DuplicateReg) {
+  const auto UseInstIter =
+      make_early_inc_range(MRI.use_instructions(DuplicateReg));
```

Comment:
Nit: I don't think we need `make_early_inc_range` here

Reply:
We need it, else the iteration below it does not work for all the uses. Without `make_early_inc_range` it will work for only the very first use.

Reply:
Trying to understand, the only thing I see that could invalidate the iterator is the Observer?

Reply:
The reason is `MO.setReg` this calls `removeRegOperandFromUseList` which updates `VRegInfo` and we are using `MRI.use_instructions` which also relies in `VRegInfo`

Reply:
But I think the `MI` is still valid and not re-allocated. Did you try without `make_early_inc_range`? I would really expect it to work

Reply:
I did try without `make_early_inc_range` it only updates the first use MI.